# Student Performance Indicator

#### Life cycle of Machine Learning Project

- Understanding the Problem Statement
- Data Collection
- Data checks to perform
- Exploratory Data Analysis
- Data Pre-Processing
- Model Training
- Choosing the Best Model

### 1) Problem Statement

- This project understands how the performance of students in exam (test scores) is affected by other variables such as Gender, race/ethinicity, Parental level of education, Type of lunch programme, Test preparation course and Scores in Math, reading and writting.

### 2) Data Collection

- Data Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
- The Dataset consists of 8 features and 1000 data points.

### 2.1 Import required packages and Dataset

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv('data/StudentsPerformance.csv')

### Show top 5 data

In [4]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

### Shape of the Dataset

In [5]:
df.shape

(1000, 8)

### 2.2 Dataset Information

- gender : sex of students  -> (Male/female)
- race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)
- parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
- lunch : having lunch before test (standard or free/reduced) 
- test preparation course : complete or not complete before test
- math score
- reading score
- writing score

### 3) Data checks to perform

- Check Missing Values
- Check Duplicates
- Check Datatypes
- Check number of unique values present in each column
- Check Statistics of the data
- Check number of categorical and numeric features present in the dataset
- Check various categories present in the different categorical variables

### 3.1 Checking Missing Values

In [8]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

#### Observation: There are no missing values in the dataset.

### 3.2 Checking Duplicates

In [9]:
df.duplicated().sum()

0

#### Observation: There are no duplicates in the dataset.

### 3.3 Check DataTypes

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


### 3.4 Checking the number of unique values of each columns

In [12]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

### 3.5 Checking Statistics of the data

In [13]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

In [14]:
df.describe(include='object')

gender race/ethnicity parental level of education     lunch  \
count     1000           1000                        1000      1000   
unique       2              5                           6         2   
top     female        group C                some college  standard   
freq       518            319                         226       645   

       test preparation course  
count                     1000  
unique                       2  
top                       none  
freq                       642

#### Insights:

- All numerical variables mean and standard deviation are closer.
- While there is a minimum score  0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17
- The number of females are slightly higher than male.
- Most of the students (642) doesn't complete the test preparation course.
- 645 students are in the regular lunch plan.
- Most of the students parents were studies in some college.